tawazi/ops.py

```diff
@@ -53,40 +64,107 @@ def to_dag(
-        it is best to construct your dag once and then consume it as much as you like in multiple threads.
+        it is best to construct your dag once and then use it as much as you like.
     Please check the example in the README for a guide to the usage.
     Args:
         declare_dag_function: a function that contains the execution of the DAG.
-        if the functions are decorated with @op decorator they can be executed in parallel.
-        Otherwise, they will be executed immediately. Currently mixing the two behaviors isn't supported.
-        max_concurrency: the maximum number of concurrent threads to execute in parallel
-        behavior: the behavior of the dag when an error occurs during the execution of a function (ExecNode)
+        Currently Only @op decorated functions can be used inside the decorated function (i.e. declare_dag_function).
+        However, you can use some simple python code to generate constants.
+        max_concurrency: the maximum number of concurrent threads to execute in parallel.
+        behavior: the behavior of the DAG when an error occurs during the execution of a function (ExecNode).
     Returns: a DAG instance
-
     """
 
-    def intermediate_wrapper(_func: Callable[..., Any]) -> Callable[..., DAG]:
-        def wrapped(*args: Any, **kwargs: Any) -> DAG:
-            # TODO: modify this horrible pattern
-            with exec_nodes_lock:
-                node.exec_nodes = []
-                _func(*args, **kwargs)
+    # wrapper used to support parametrized and non parametrized decorators
+    def intermediate_wrapper(_func: Callable[..., Any]) -> DAG:
+
+        # 0. Protect against multiple threads declaring many DAGs at the same time
+        with exec_nodes_lock:
+
+            # 1. node.exec_nodes contains all the ExecNodes that concern the DAG being built at the moment.
+            #      make sure it is empty
+            node.exec_nodes = []
+            try:
+                # 2. make ExecNodes corresponding to the arguments of the ExecNode
+                # 2.1 get the names of the arguments and the default values
+                func_args, func_default_args = get_args_and_default_args(_func)
+
+                # 2.2 Construct non default arguments.
+                # Corresponding values must be provided during usage
+                args: List[ExecNode] = [ArgExecNode(_func, arg_name) for arg_name in func_args]
+                # 2.2 Construct Default arguments.
+                args.extend(
+                    [
+                        ArgExecNode(_func, arg_name, arg)
+                        for arg_name, arg in func_default_args.items()
+                    ]
+                )
+                # 2.3 Arguments are also ExecNodes that get executed inside the scheduler
+                node.exec_nodes.extend(args)
+
+                # 3. Execute the dependency describer function
+                # NOTE: Only ordered parameters are supported at the moment!
```

Comment:
do you think we should wait for the next release to support them ?

Reply:
I think we can do a usable release and then introduce further improvements in the minor version. The goal is to start using `tawazi~=0.3.0` ASAP

Reply:
so let's wait for the next release and stop adding features in this one